README.md
# FGS Buddy :sunglasses:


<div style="text-align: center;">

![Buddyn](Buddy.ico)<br>



En grym liten FGS-paketerare för att skapa SIP-paket till e-arkiv.
</div>

---


![](screenv1_1.PNG)


## Features :star:
* Följer Riksarkivets standard FGS-paketstruktur version 1.2.
* Kan hantera innehåll (content) som är lagrat i undermappar i flera nivåer.  
* Genererar checksummor, identifierar MIME-typer, filstorlekar och tidpunkt för skapande för de filer som ingår i SIP-paketet.
* Ger filerna korrekta filnamn enligt FGS-standarden samtidigt som originalnamnet bevaras som metadata.   
* Ger möjlighet att lägga till schemafil och metadatafil i paketet. 
* Innnehåller tooltips med besrivande hjälptexter hämtade från standarden för alla fält + hänvisning till var fältet hamnar i sip.xml.
* Sparar paketet som en zip-fil som namnges "Arkivbildarnamn_Systemnamn_ÅÅÅÅ_MM_DDTHH_MM_SS.zip". 

---

## Struktur på SIP-paketet :package:

Exempel på strukturen i ett enkelt SIP-paket skapat av FGS-Buddy: <br>

![](tree.PNG)

---

## Kom igång :rocket:

1. [Ladda ner den senaste releasen av FGS-Buddy.](https://github.com/Viktor-Lundberg/FGSBuddy/releases)
1. Packa upp zip-filen på din dator där du vill lagra applikationen.
1. Starta FGS-Buddy med "FGSbuddyGUI.exe".
1. Fyll i relevant metadata för ditt SIP-paket.
1. Välj vilken katalog som paketinnehållet (filerna) ska hämtas ifrån och om även underkataloger ska ingå i paketet.
1. Välj eventuell metadatafil och schemafil.
1. Välj output-katalog där Buddyn ska spara SIP-paketet.
1. Klicka på "Skapa paket" - Buddyn ger feedback ifall något av de obligatoriska värdena saknas, annars genereras SIP-paketet.
1. Hämta upp ditt färdiga SIP-paket i vald katalog.   

---

## Kända problem :warning:

* Applikationen kan ta lång tid att starta om den körs från en nätverksdisk. Rekommendationen är att alltid installera FGS-buddy på en lokal disk (exempelvis c:).
* Tooltips på flera rader laddas inte alltid. - workaround flytta muspekaren från fältet och försök igen.



---

## Credits :trophy:

* Martin Olsson - Har bidragit med funktionstester, idéer kring funktionalitet och feedback kring implementering av FGS-paketstruktur.
* Dan Lundh - Agerat bollplank kring FGS-buddy.
* Glasögon till ikonen hämtade från - http://clipart-library.com/clipart/1943208.htm  

---
<br>
<a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/">Creative Commons Attribution-NonCommercial 4.0 International License</a>.



releasenotes.txt
Version 1.1* Möjlighet att välja outputkatalog för SIP-paketet.* Meny som ger direktåtkomst till FGS-dokumentationen.* Möjlighet att kunna ange metadata om "Avtalsform" för SIP-paketet.* Avsnittet "Om FGS-Buddy" med bland annat release notes.* Funktionalitet att köra applikationen i helskärmsläge/valfri storlek.* Fixat bugg där applikationen kraschade när man fyllde i "informationsägare".* Knapp för att enkelt rensa alla värden.* Förbättringar i gränssnittet.Version 1.0* Följer Riksarkivets standard FGS-paketstruktur version 1.2.* Kan hantera innehåll (content) som är lagrat i undermappar i flera nivåer.* Genererar checksummor, identifierar MIME-typer, filstorlekar och tidpunkt för skapande för de filer som ingår i SIP-paketet.
* Ger filerna korrekta filnamn enligt FGS-standarden samtidigt som originalnamnet bevaras som metadata.
* Ger möjlighet att lägga till schemafil och metadatafil i paketet.
* Innnehåller tooltips med besrivande hjälptexter hämtade från standarden för alla fält + hänvisning till var fältet hamnar i sip.xml.
* Sparar paketet som en zip-fil som namnges "Arkivbildarnamn_Systemnamn_ÅÅÅÅ_MM_DDTHH_MM_SS.zip".

FGSbuddyGUI.py
import os
import time
import PySimpleGUI as sg
import fgsbuddyfunctions as FGSfunc
import json
import webbrowser

# Funktion för att rensa alla input-fält
def clearinput():
    saveinput = ['folderinput', 'outputfolderinput']
    # Går igenom key_dict och kontrollerar vad det är för sg objekt
    for k, v in window.key_dict.items():
        # Om fältet är ett inputfält och keyn inte finns med i listan över värden som inte ska ändras --> Rensa 
        if isinstance(v, sg.Input) and k not in saveinput:
            v.update('')
        # Sätter alla combofält till defaultvärdet
        elif isinstance(v, sg.Combo):
            v.update(v.DefaultValue)
        # Ändrar till defaultvärdet i radioknapparna kring om paketet ska innehålla "undermappar"
        elif isinstance(v, sg.Radio) and k == 'subfoldersfalse':
            v.update(True)
        else:
            continue
    window.refresh()


# Funktion för att skapa popup-fönstret "Om FGS-Buddy"
def buddywindow():
    # Hämtar alla releasenotes från txt-fil och lägger i variabeln releasenotes
    with open('releasenotes.txt', encoding='utf-8') as releasenotesdoc:
        releasenotes = releasenotesdoc.read()

    # Layout för fönstret
    aboutlayout = [
        [sg.Text('FGS-Buddy', font='Arial 12 bold', size=30)],
        [sg.Text(f'Version: {version}\nSkapad av: Viktor Lundberg\nLicens: Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)', font='Consolas 10')],
        [sg.Text('FGS-Buddy på github',font='Consolas 10 underline', text_color='blue', enable_events=True, key='githublink')],
        [sg.Text('', font='Arial 12 bold', size=30)],
        [sg.Text('Release notes', font='Arial 10 bold', size=30)],
        [sg.Text(releasenotes, font='Consolas 8')],
        [sg.Button('Ok', key='Exit', size=7)]
    ]

    # Skapar nya fönstret
    about = sg.Window('Om FGS-Buddy', aboutlayout, icon="Buddy.ico")
    
    # Programloop för "Om FGS-Buddy"-fönstret.
    while True:
        event, values = about.read()
        if event == "Exit" or event == sg.WIN_CLOSED:
            about.close()
            releasenotesdoc.close()
            break
        if event == 'githublink':
            webbrowser.open('https://github.com/Viktor-Lundberg/FGSBuddy')


# Nuvarande version
version = '1.1.0'
# Current working directory
cwd = os.getcwd()


# Laddar tooltipsfil
tooltipsfile = open('tooltips.json', encoding='utf-8')
tt = json.load(tooltipsfile)
# Fixar jsonformateringen i tooltips \\\n --> \n
for k, v in tt.items():
    tt[k] = v.replace('\\', '')


# Sätter färgtema
sg.theme('greenMono') #LightGreen2 DarkBlue3 #reddit greenMono

# Listor med element som utgör olika delar av layouten. 

paketinformation = [
    [sg.Text('Övergripande paketinformation', font='Arial 12 bold',size=30)], 
    [sg.Text('Status*', size=32, tooltip=tt['StatusT']), sg.Combo(['NEW', 'SUPPLEMENT', 'REPLACEMENT','TEST','VERSION', 'OTHER'], default_value='NEW', key='recordstatus', tooltip=tt['Status'], size=43) ],
    [sg.Text('Beskrivning', tooltip=tt['BeskrivningT'], size=32),sg.Input(key='beskrivning', tooltip=tt['Beskrivning'])],
    [sg.Text('Leveransöverenskommelse*', size=32, tooltip=tt['LeveransoverenskommelseT']), sg.Input(key='submissionagreement',tooltip=tt['Leveransoverenskommelse'])],
    [sg.Text('Tidigare leveransöverenskommelse', size=32, tooltip=tt['TidigareLeveransoverenskommelseT']),sg.Input(key='formersubmissionagreement', tooltip=tt['TidigareLeveransoverenskommelse'])],
    [(sg.Text('Överföring', size=32, tooltip=tt['OverforingT'])), sg.Input(key='overforing', tooltip=tt['Overforing'])],
    [sg.Text('Ordningsnummer inom överföring', size=32, tooltip=tt['OrdningsnummerT']), sg.Input(key='overforingNR',tooltip=tt['Ordningsnummer'])]

]

parter = [
    [sg.Text('Arkivbildare', font='Arial 12 bold', size=25)],    
    [sg.Text('Arkivbildare - Namn*', tooltip=tt['ArkivbildarenamnT']) ],
    [sg.Input(key='arkivbildare', tooltip=tt['Arkivbildarenamn'])],
    [sg.Text('Arkivbildare - Identitetskod*', tooltip=tt['ArkivbildareIDkodT'])],
    [sg.Combo(['VAT', 'DUNS', 'ORG','HSA','Local', 'URI'], default_value='ORG', key='IDkodtyp'),sg.Input(key='IDkod',size=35, tooltip=tt['ArkivbildareIDkod'])],
    [sg.Text('Levererande organisation - Namn*', tooltip=tt['LevererandeorganisationnamnT']) ],
    [sg.Input(key='levererandeorganisation', tooltip=tt['Levererandeorganisationnamn'])],
    [sg.Text('Levererande organisation - Identitetskod', tooltip=tt['LevererandeorganisationIDkodT'])],
    [sg.Combo(['VAT', 'DUNS', 'ORG','HSA','Local', 'URI'], default_value='ORG', key='IDkodtyplevererandeorganisation'),sg.Input(key='IDkodlevererandeorganisation',size=35, tooltip=tt['LevererandeorganisationIDkod'])],
    
    ]

system = [
    [sg.Text('Källsystem', font='Arial 12 bold', size=30)], 
    [sg.Text('Systemnamn*', tooltip=tt['SystemnamnT'])],
    [sg.Input(key='system', tooltip=tt['Systemnamn'])],
    [sg.Text('Systemversion',tooltip=tt['SystemversionT'])],
    [sg.Input(key='systemversion', tooltip=tt['Systemversion'])],
    [sg.Text('Systemtyp', tooltip=tt['SystemtypT'])],
    [sg.Input(key='systemtyp', tooltip=tt['Systemtyp'])],

    ]

innehall = [  
    [sg.Text('Innehåll', font='Arial 12 bold', size=30)],
    [sg.Text('Sökväg till filer som ska ingå i paketet')],
    [sg.Input(default_text=cwd,tooltip="Välj katalog", key='folderinput'), sg.FolderBrowse('Välj katalog',key="folder", initial_folder=os.path.join(cwd))],
    [sg.Text('Inkludera undermappar')],
    [sg.Radio('Ja', 'subfolders', default=False, key='subfolderstrue'), sg.Radio('Nej', 'subfolders', default=True, key='subfoldersfalse')],
    [sg.Text('Sökväg till metadatafil')],
    [sg.Input(tooltip="Välj metadatafil", key='metadatafile'), sg.FileBrowse('Välj fil',key="metadatafileB", initial_folder=os.path.join(cwd))],
    [sg.Text('Sökväg till schemafil')],
    [sg.Input(tooltip="Välj schemafil", key='schemafile'), sg.FileBrowse('Välj fil', key="schemafileB", initial_folder=os.path.join(cwd) )],

    ]

information = [
    [sg.Text('Information', font='Arial 12 bold', size=30)],
    [sg.Text('Informationstyp*', size=32, tooltip=tt['InformationstypT']), sg.Combo(['ERMS','Personnel','Medical record','Economics','Databases','Webpages', 'GIS', 'No specification', 'AIC', 'Archival information','Unstructured', 'Single records', 'Publication'], default_value='ERMS', key='informationstyp', tooltip=tt['Informationstyp'], size=43)],
    [sg.Text('Informationstypsspecifikation', size=32, tooltip=tt['InformationstypsspecifikationT']),sg.Input(key='informationstypsspecifikation', tooltip=tt['Informationstypsspecifikation'])],
    [sg.Text('Tidsomfång', size=32), sg.CalendarButton('Startdatum', target='startdatum', format='%Y-%m-%d', tooltip=tt['StartdatumT']), sg.Input(key='startdatum', size=(10,1), tooltip=tt['Startdatum']), sg.Text('-'), sg.Input(key='slutdatum', size=(10,1), tooltip=tt['Slutdatum']),sg.CalendarButton('Slutdatum', target='slutdatum', format='%Y-%m-%d', tooltip=tt['SlutdatumT'])],
    [sg.Text('Informationsklass', size=32, tooltip=tt['InformationsklassT']), sg.Input(key='informationsklass', tooltip=tt['Informationsklass'])],
    [sg.Text('Sekretess', size=32, tooltip=tt['SekretessT']), sg.Combo(['Secrecy', 'PuL', 'Secrecy and PuL','GDPR','Not specified'], default_value='Not specified', key='sekretess', tooltip=tt['Sekretess'], size=43)],
    [sg.Text('Gallring', size=32, tooltip=tt['GallringT']),sg.Combo(['Yes', 'No'], default_value='No', key='gallring', tooltip=tt['Gallring'], size=43) ],
]

space = [
    [sg.Text('', font='Arial 8 bold', size=30)],
    ]

ovrigt = [
    [sg.Text('Övrig information', font='Arial 12 bold', size=30)],
    [(sg.Text('Arkivets namn',size=32, tooltip=tt['ArkivetsnamnT'])), sg.Input(key='arkivetsnamn', tooltip=tt['Arkivetsnamn'])],
    [(sg.Text('Arkivets referenskod',size=32, tooltip=tt['ArkivetsreferenskodT'])), sg.Input(key='arkivetsreferenskod',tooltip=tt['Arkivetsreferenskod'])],
    [(sg.Text('Tidigare referenskod',size=32, tooltip=tt['TidigareReferenskodT'])),sg.Input(key='tidigarereferenskod', tooltip=tt["TidigareReferenskod"])],
    [(sg.Text('Producerande organisation - Namn',size=32, tooltip=tt['ProducerandeorganisationnamnT'])), sg.Input(key='prodorgnamn', tooltip=tt['Producerandeorganisationnamn'])],
    [(sg.Text('Producerande organisation- IDkod',size=32, tooltip=tt['ProducerandeorganisationIDkodT'])),sg.Combo(['VAT', 'DUNS', 'ORG','HSA','Local', 'URI'], default_value='ORG', key='prodorgIDkodtyp'),sg.Input('', key='prodorgIDkod',size=35, tooltip=tt['ProducerandeorganisationIDkod'])],
    [(sg.Text('Avsändande organisation - Namn',size=32, tooltip=tt['AvsandandeorganisationnamnT'])), sg.Input(key='avsandandeorgnamn', tooltip=tt['Avsandandeorganisationnamn'])],
    [(sg.Text('Avsändande organisation - IDkod',size=32, tooltip=tt['AvsandandeorganisationIDkodT'])),sg.Combo(['VAT', 'DUNS', 'ORG','HSA','Local', 'URI'], default_value='ORG', key='avsandandeorgIDkodtyp'),sg.Input('', key='avsandandeorgIDkod',size=35, tooltip=tt['AvsandandeorganisationIDkod'])],
    [(sg.Text('Informationsägande org. - Namn',size=32, tooltip=tt['InfoagarenamnT'])), sg.Input(key='infoagarenamn', tooltip=tt['Infoagarenamn'])],
    [(sg.Text('Informationsägande org. - IDkod',size=32, tooltip=tt['InfoagareIDkodT'])),sg.Combo(['VAT', 'DUNS', 'ORG','HSA','Local', 'URI'], default_value='ORG', key='infoagareIDkodtyp'),sg.Input('', key='infoagareIDkod',size=35)],
]

ovrigt2 = [
    [sg.Text('', font='Consolas 12', size=30)],
    [(sg.Text('Levererande system - Namn',size=32, tooltip=tt['LevererandesystemnamnT'])), sg.Input(key='levererandesystemnamn', tooltip=tt['Levererandesystemnamn'])],
    [(sg.Text('Levererande system - Version',size=32, tooltip=tt['LevererandesystemversionT'])), sg.Input(key='levererandesystemversion', tooltip=tt['Levererandesystemversion'])],
    [(sg.Text('Konsult - Namn',size=32, tooltip=tt['KonsultnamnT'])), sg.Input(key='konsultnamn', tooltip=tt['Konsultnamn'])],
    [(sg.Text('Konsult - Identitetskod',size=32,tooltip=tt['KonsultIDkodT'] )),sg.Combo(['VAT', 'DUNS', 'ORG','HSA','Local', 'URI'], default_value='ORG', key='konsultIDkodtyp'),sg.Input('', key='konsultIDkod',size=35, tooltip=tt['KonsultIDkod'])],
    [(sg.Text('Mottagare - Namn',size=32, tooltip=tt['MottagarenamnT'])), sg.Input(key='mottagarenamn', tooltip=tt['Mottagarenamn'])],
    [(sg.Text('Mottagare - Identitetskod',size=32, tooltip=tt['MottagareIDkodT'])),sg.Combo(['VAT', 'DUNS', 'ORG','HSA','Local', 'URI'], default_value='ORG', key='mottagareIDkodtyp'),sg.Input('', key='mottagareIDkod',size=35, tooltip=tt['MottagareIDkod'])],
    [(sg.Text('Kontaktperson - Namn',size=32, tooltip=tt['KontaktpersonnamnT'])), sg.Input(key='kontaktpersonnamn', tooltip=tt['Kontaktpersonnamn'])],
    [(sg.Text('Kontaktperson - Kontaktuppgifter',size=32, tooltip=tt['KontaktuppgifterT'])), sg.Input(key='kontaktuppgifter', tooltip=tt['Kontaktuppgifter'])],
    [sg.Text('Avtalsform', size=32, tooltip=tt['AvtalsformT']), sg.Combo(['AGREEMENT', 'DEPOSIT', 'GIFT','Not specified'], default_value='Not specified', key='avtalsform', tooltip=tt['Avtalsform'], size=43)],

]

# Menyraden
meny = [
    ['FGS-dokumentation', ['FGS-Paketstruktur v 1.2','FGS-Paketstruktur v 1.2 - tillägg','Schema v 1.2']],['Hjälp', ['Om FGS-Buddy']]
     ]

# GUI layout
layout = [
    # OBS! Pysimplegui har problem med custom menubar (om det används syns inte applikationen i verktygsfältet, använd classic tills fix...)
    #[sg.Titlebar('FGS-Buddy v 1.1 - Viktor Lundberg', font='Consolas 10', background_color='Black')],
    #[sg.MenubarCustom(meny, bar_background_color='Pink', bar_text_color='Black')],
    
    [sg.MenuBar(meny, background_color='Pink')],
    [sg.Column(paketinformation, vertical_alignment='top'), sg.Column(information)],
    [sg.Column(space)],
    [sg.Column(parter, vertical_alignment='top'), sg.Column(system, vertical_alignment='top'), sg.Column(innehall, vertical_alignment='top')],
    [sg.pin(sg.Column(ovrigt, key='invisible', visible=False)), sg.Column(ovrigt2, key='invisible2', visible=False, vertical_alignment='top')],
    [sg.Text('')],
    [sg.Output(size=(165,5), key='output', pad=5, background_color=	'pink', echo_stdout_stderr=True)],
    [sg.Text('Outputkatalog'),sg.Input(default_text=cwd, tooltip="Välj katalog", size=65, key='outputfolderinput'), sg.FolderBrowse('Välj katalog',key="outputfolder", initial_folder=os.path.join(cwd)),sg.Submit('Skapa paket', key='createSIP', size=15,button_color='black on pink'),sg.Text('', size=15), sg.Button('Rensa', key='clear', size=15),sg.Button('Visa alla fält', key='fields', size=15)],
    ]



# Skapar "huvudfönstret"
window = sg.Window(f'FGS-Buddy v {version}',layout, font='Consolas 10', icon="Buddy.ico", resizable=True, titlebar_background_color='green') 


# Variabler för att kontrollera obligatoriska värden samt trigger för att visa/dölja alla element i layouten.
forcedvaluesdict = {}
allvalues = False


# Program-Loop
while True:
    event, values = window.read()
    match event:
        case sg.WIN_CLOSED:
            break

        # Startar processen för att skapa FGS-paket om användaren trycker på "skapa paket"-knappen
        case 'createSIP':
            window.find_element('output').Update('')
            window.refresh()
            metadatafile = False
            schemafile = False
            subfolders = False

            # Kontrollerar att alla obligatoriska värden har information!
            if values['submissionagreement'] == '' or values['system'] == '' or values['arkivbildare'] == '' or values['IDkod'] == '' or values['levererandeorganisation'] == '':
                # Lägger till värden i dicten för att kunna visa användaren vilka värden som saknas
                forcedvaluesdict['System'] = values['system']
                forcedvaluesdict['Leveransöverenskommelse'] = values['submissionagreement']
                forcedvaluesdict['Arkivbildare'] = values['arkivbildare']
                forcedvaluesdict['Identitetskod'] = values['IDkod']
                forcedvaluesdict['Levererande organisation'] = values['levererandeorganisation']
                # Loopar igenom dicten
                for k, v in forcedvaluesdict.items():
                    # Returnerar de värden som saknas till användaren.
                    if v == '':
                        print(f'Fältet "{k}" saknar värde.')
                
            # Om allt är ok skapa paketet med inkommande parametrar.
            else:
                print(f'Skapar FGS-paketet...')
                window.refresh()
                fgsPackage = FGSfunc.FgsMaker(values)
                if values['folder'] == '':
                    folder = os.path.join(cwd)
                else:
                    folder = os.path.join(values['folder'])
                if values['outputfolder'] == '':
                    outputfolder = os.path.join(cwd)
                else:
                    outputfolder = os.path.join(values['outputfolder'])
                if values['subfolderstrue']:
                    subfolders = True
                if values['metadatafile'] != '':
                    metadatafile = values['metadatafile']
                if values['schemafile'] != '':
                    schemafile = values['schemafile']
                fgsPackage.collectFiles(folder, subfolders, metadatafile, schemafile)
                fgsPackage.createSip()
                fgsPackage.createFgsPackage(cwd,outputfolder)
                time.sleep(0.3)
                window.find_element('output').update('')
                window.refresh()
                print(fgsPackage.output)
        
        # Visar eller döljer layout om användaren trycker på knappen
        case 'fields':
            if allvalues == False:
                window['invisible'].Update(visible=True)
                window['invisible2'].Update(visible=True)
                window.refresh()
                window['fields'].Update('Dölj fält')
                allvalues = True
            else:
                window['invisible'].Update(visible=False)
                window['invisible2'].Update(visible=False)
                window.refresh()
                window['fields'].Update('Visa alla fält')
                allvalues = False
        
        case 'clear':
            clearinput()
        
        # Meny - FGS-dokumentation
        case 'FGS-Paketstruktur v 1.2':
            webbrowser.open('https://riksarkivet.se/Media/pdf-filer/doi-t/FGS_Paketstruktur_RAFGS1V1_2.pdf')
        case 'FGS-Paketstruktur v 1.2 - tillägg':
            webbrowser.open('https://riksarkivet.se/Media/pdf-filer/doi-t/FGS_Paketstruktur_Tillagg_RAFGS1V1_2A20171025.pdf')
        case 'Schema v 1.2':
            webbrowser.open('http://xml.ra.se/e-arkiv/METS/CSPackageMETS.xsd')
        
        # Meny - Hjälp
        case 'Om FGS-Buddy':
            sg.Window.disappear(window)
            buddywindow()
            sg.Window.reappear(window)